### ISSUE TITLE
Inconsistent labeling on envoy cluster metrics

### ISSUE THREAD

**Issue by Stono** [istio/istio#51761]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

I wanted to get envoy metrics around retries, so I added:

```
        proxy.istio.io/config: |-
          proxyStatsMatcher:
            inclusionRegexps:
            - ".*upstream_rq_retry.*"
```

The clusters on that workload are:

```
SERVICE FQDN                                          PORT     SUBSET     DIRECTION     TYPE             DESTINATION RULE
                                                      8080     -          inbound       ORIGINAL_DST     unified-registration-system.unified-registration-system
                                                      8081     -          inbound       ORIGINAL_DST     
BlackHoleCluster                                      -        -          -             STATIC           
InboundPassthroughClusterIpv4                         -        -          -             ORIGINAL_DST     
PassthroughCluster                                    -        -          -             ORIGINAL_DST     
accounts.google.com                                   443      -          outbound      LOGICAL_DNS      
agent                                                 -        -          -             STATIC           
api.clickatell.com                                    80       -          outbound      LOGICAL_DNS      clickatel.unified-registration-system
api.clickatell.com                                    443      -          outbound      LOGICAL_DNS      clickatel.unified-registration-system
api.textmarketer.co.uk                                80       -          outbound      LOGICAL_DNS      textmarketer.unified-registration-system
api.textmarketer.co.uk                                443      -          outbound      LOGICAL_DNS      textmarketer.unified-registration-system
api.voxanalytics.com                                  80       -          outbound      LOGICAL_DNS      iovox.unified-registration-system
api.voxanalytics.com                                  444      -          outbound      LOGICAL_DNS      iovox.unified-registration-system
app.apm.svc.cluster.local                             80       -          outbound      EDS              apm.apm
app.platform-identity-provider.svc.cluster.local      80       -          outbound      EDS              platform-identity-provider.platform-identity-provider
app.platform-smtp-proxy.svc.cluster.local             80       -          outbound      EDS              platform-smtp-proxy.platform-smtp-proxy
app.unified-registration-system.svc.cluster.local     80       -          outbound      EDS              unified-registration-system.unified-registration-system
appleid.apple.com                                     80       -          outbound      LOGICAL_DNS      apple.unified-registration-system
appleid.apple.com                                     443      -          outbound      LOGICAL_DNS      apple.unified-registration-system
graph.facebook.com                                    80       -          outbound      LOGICAL_DNS      facebook.unified-registration-system
graph.facebook.com                                    443      -          outbound      LOGICAL_DNS      facebook.unified-registration-system
kafka-prom-exporter.kafka.svc.cluster.local           80       -          outbound      EDS              kafka-prom-exporter.kafka
kube-dns.kube-system.svc.cluster.local                53       -          outbound      EDS              
limited.facebook.com                                  80       -          outbound      LOGICAL_DNS      facebook-limited.unified-registration-system
limited.facebook.com                                  443      -          outbound      LOGICAL_DNS      facebook-limited.unified-registration-system
logstash.core-system.svc.cluster.local                9198     -          outbound      EDS              
logstash.core-system.svc.cluster.local                9600     -          outbound      EDS              
metadata.google.internal                              80       -          outbound      EDS              
oauth2.googleapis.com                                 443      -          outbound      LOGICAL_DNS      
prometheus_stats                                      -        -          -             STATIC           
schema-registry.kafka.svc.cluster.local               80       -          outbound      EDS              schema-registry.kafka
sds-grpc                                              -        -          -             STATIC           
smtp.platform-smtp-proxy.svc.cluster.local            25       -          outbound      EDS              smtp.platform-smtp-proxy
sqladmin.googleapis.com                               443      -          outbound      LOGICAL_DNS      
storage.googleapis.com                                80       -          outbound      LOGICAL_DNS      googleapis-storage.unified-registration-system
storage.googleapis.com                                443      -          outbound      LOGICAL_DNS      googleapis-storage.unified-registration-system
www.googleapis.com                                    443      -          outbound      LOGICAL_DNS      
xds-grpc                                              -        -          -             STATIC           
zipkin                                                -        -          -             STRICT_DNS       
zipkin.istio-system.svc.cluster.local                 9411     -          outbound      EDS              disable-mtls-to-zipkin.istio-system
```

Notice how for some hosts we're using tls origination.  This gives a real miss-match of metrics, so for all the "simple" hosts we get:

```
# TYPE envoy_cluster_upstream_rq_retry counter
envoy_cluster_upstream_rq_retry{cluster_name="BlackHoleCluster"} 0
envoy_cluster_upstream_rq_retry{cluster_name="InboundPassthroughClusterIpv4"} 0
envoy_cluster_upstream_rq_retry{cluster_name="PassthroughCluster"} 0
envoy_cluster_upstream_rq_retry{cluster_name="agent"} 0
envoy_cluster_upstream_rq_retry{cluster_name="inbound|8080||"} 0
envoy_cluster_upstream_rq_retry{cluster_name="inbound|8081||"} 0
envoy_cluster_upstream_rq_retry{cluster_name="outbound|25||smtp.platform-smtp-proxy.svc.cluster.local"} 0
envoy_cluster_upstream_rq_retry{cluster_name="outbound|53||kube-dns.kube-system.svc.cluster.local"} 0
envoy_cluster_upstream_rq_retry{cluster_name="outbound|80||app.apm.svc.cluster.local"} 0
envoy_cluster_upstream_rq_retry{cluster_name="outbound|80||app.platform-identity-provider.svc.cluster.local"} 0
envoy_cluster_upstream_rq_retry{cluster_name="outbound|80||app.platform-smtp-proxy.svc.cluster.local"} 0
envoy_cluster_upstream_rq_retry{cluster_name="outbound|80||app.unified-registration-system.svc.cluster.local"} 0
envoy_cluster_upstream_rq_retry{cluster_name="outbound|80||kafka-prom-exporter.kafka.svc.cluster.local"} 0
envoy_cluster_upstream_rq_retry{cluster_name="outbound|80||schema-registry.kafka.svc.cluster.local"} 0
envoy_cluster_upstream_rq_retry{cluster_name="outbound|9198||logstash.core-system.svc.cluster.local"} 0
envoy_cluster_upstream_rq_retry{cluster_name="outbound|9411||zipkin.istio-system.svc.cluster.local"} 0
envoy_cluster_upstream_rq_retry{cluster_name="outbound|9600||logstash.core-system.svc.cluster.local"} 0
envoy_cluster_upstream_rq_retry{cluster_name="prometheus_stats"} 0
envoy_cluster_upstream_rq_retry{cluster_name="sds-grpc"} 0
envoy_cluster_upstream_rq_retry{cluster_name="xds-grpc"} 0
envoy_cluster_upstream_rq_retry{cluster_name="zipkin"} 0
```

Which is what I'd expect (a singular `envoy_cluster_upstream_rq_retry` with `cluster_name` as a label), however for tls origination we get:

```
# TYPE envoy_cluster_voxanalytics_com_upstream_rq_retry counter
envoy_cluster_voxanalytics_com_upstream_rq_retry{cluster_name="outbound|444||api"} 0
envoy_cluster_voxanalytics_com_upstream_rq_retry{cluster_name="outbound|80||api"} 0
```

eg the cluster name is no longer a label, and instead it has become part of the metric name.  The `cluster_name` in the `config_dump` is `outbound|80||api.voxanalytics.com`

I would expect `envoy_cluster_upstream_rq_retry` to be the correct one with `cluster_name` as a label, so feels like this is broken for tls origination. 

In case it helps, here is a `VirtualService` for one of those hosts. 

```
apiVersion: networking.istio.io/v1
kind: VirtualService
metadata:
  name: iovox
  namespace: unified-registration-system
spec:
  exportTo:
  - .
  hosts:
  - api.voxanalytics.com
  http:
  - match:
    - port: 80
    name: httpPort
    retries:
      attempts: 3
      perTryTimeout: 10s
      retryOn: connect-failure,refused-stream,unavailable,cancelled,resource-exhausted,502,503
    route:
    - destination:
        host: api.voxanalytics.com
        port:
          number: 80
      headers:
        request:
          remove:
          - x-envoy-attempt-count
          - x-envoy-decorator-operation
          - x-envoy-expected-rq-timeout-ms
          - x-envoy-hedge-on-per-try-timeout
          - x-envoy-is-timeout-retry
          - x-envoy-max-retries
          - x-envoy-original-path
          - x-envoy-peer-metadata
          - x-envoy-peer-metadata-id
          - x-envoy-retriable-header-names
          - x-envoy-retriable-status-codes
          - x-envoy-retry-grpc-on
          - x-envoy-retry-on
          - x-envoy-upstream-alt-stat-name
          - x-envoy-upstream-rq-per-try-timeout-ms
          - x-envoy-upstream-rq-timeout-alt-response
          - x-envoy-upstream-rq-timeout-ms
          - x-envoy-upstream-service-time
          - x-envoy-upstream-stream-duration-ms
          set:
            host: api.voxanalytics.com
      weight: 100
    timeout: 10s
  - match:
    - port: 444
    name: httpsPort
    retries:
      attempts: 3
      perTryTimeout: 10s
      retryOn: connect-failure,refused-stream,unavailable,cancelled,resource-exhausted,502,503
    route:
    - destination:
        host: api.voxanalytics.com
        port:
          number: 444
      headers:
        request:
          remove:
          - x-envoy-attempt-count
          - x-envoy-decorator-operation
          - x-envoy-expected-rq-timeout-ms
          - x-envoy-hedge-on-per-try-timeout
          - x-envoy-is-timeout-retry
          - x-envoy-max-retries
          - x-envoy-original-path
          - x-envoy-peer-metadata
          - x-envoy-peer-metadata-id
          - x-envoy-retriable-header-names
          - x-envoy-retriable-status-codes
          - x-envoy-retry-grpc-on
          - x-envoy-retry-on
          - x-envoy-upstream-alt-stat-name
          - x-envoy-upstream-rq-per-try-timeout-ms
          - x-envoy-upstream-rq-timeout-alt-response
          - x-envoy-upstream-rq-timeout-ms
          - x-envoy-upstream-service-time
          - x-envoy-upstream-stream-duration-ms
          set:
            host: api.voxanalytics.com
      weight: 100
    timeout: 10s
```

### Version

```Text
1.21.2
```

**Comment by Stono** [comment]
Digging a bit more, what appears to be happening here is in the metrics, the first part of the domain is being used for the cluster name

```
# TYPE envoy_cluster_facebook_com_upstream_rq_retry counter
envoy_cluster_facebook_com_upstream_rq_retry{cluster_name="outbound|443||graph"} 0
envoy_cluster_facebook_com_upstream_rq_retry{cluster_name="outbound|443||limited"} 0
envoy_cluster_facebook_com_upstream_rq_retry{cluster_name="outbound|80||graph"} 0
envoy_cluster_facebook_com_upstream_rq_retry{cluster_name="outbound|80||limited"} 0
```

For example, here the `cluster_name` is actually `outbound|443||graph.facebook.com` (or `limited.facebook.com`) in the `config_dump` but in these metrics it's being split

**Comment by Stono** [comment]
I wonder if this is because we're removing `x-envoy-retriable-status-codes` from the outbound response, is that used internally to the retry logic?

I'l try adding it back.

**Comment by keithmattix** [comment]
@zirain @howardjohn any idea where this is configured? Seems like a parsing bug due to the `.` delimeter

**Comment by keithmattix** [comment]
Ok so because envoy uses `.` as the stat delimiter, it's impossible to parse correctly. The only thing we can do is to have an alternate name for stats that replaces `.` with `_` so that stats don't break.

**Comment by zirain** [comment]
> Ok so because envoy uses `.` as the stat delimiter, it's impossible to parse correctly. The only thing we can do is to have an alternate name for stats that replaces `.` with `_` so that stats don't break.

I recall there's similar one on grpc stats, I tried to fix it in the past, but didn't find a solution.

**Comment by keithmattix** [comment]
For clusters, I was able to put together a solution using `alt_stat_name`: #51797

**Comment by keithmattix** [comment]
Fixed by #51861
